### DISCUSSION TITLE
Can I use .rgignore to list the specific folders that SHOULD be searched?

### DISCUSSION THREAD

**Discussion by professionalPeter** [BurntSushi/ripgrep#2240] [Q&A] [answered]
Is it possible to use .rgignore to ignore all folders under the current directory except paths specified in the .rgignore and have those exceptions still searched recursively?

I've gone through the documentation for .rgignore and tried quite a few variations, but have not been successful. The closest I've come is ignoring all the folders under the current directory except the specified exceptions, but without those exceptions being searched recursively. 

I'm running this under Windows 10.

**Comment by BurntSushi** [comment, marked as answer]
```
*
!/foo
```

That ignores everything except for `foo` in the same directory as the rgignore.

It's hard to tell if this is the answer you're looking for though. Namely, you've described your problem instead of *showing* your problem. It would be better to include a concrete example that you want to work.

**Reply by professionalPeter** [reply]
Thanks for the response despite me not forming the question well. 
So if I have these folders and files where file.txt always contains "foo": 

```
|-file.txt
|-a/
   |-file.txt
   |-b/
      |- file.txt
      |-c/
         |-file.txt
|-a2/
   |-file.txt
   |-b/
      |- file.txt
      |-c/
         |-file.txt
```

and put the following in my .rgignore
```
*
!/a
```

and run `rg foo --stats`, then I get
```
No files were searched, which means ripgrep probably applied a filter you didn't expect.
Running with --debug will show why files are being skipped.

0 matches
0 matched lines
0 files contained matches
0 files searched
0 bytes printed
0 bytes searched
0.000000 seconds spent searching
0.148224 seconds
```
Of course if I remove the first entry from the .rgignore, then all 7 files are searched.

**Ideally**, what I'm trying to achieve is that the root file.txt and the 3 file.txts in the "a/" hierarchy are all searched. However if I could even get only the 3 file.txts below the "a/" folder searched that would be helpful

Here's the output when run with `--debug`: [rg.txt](https://github.com/BurntSushi/ripgrep/files/8941480/rg.txt)

**Reply by BurntSushi** [reply]
Ah right okay, so you do need one extra rule:

```
$ cat .rgignore
*
!/a
!/a/**
```

And then your example works. The issue here is that `/a` whitelists `a`, but the `*` still gets applied to, e.g., `a/file.txt` and `a/b`. So you need the extra `/a/**` to say "and also whitelist anything below `a` too."

As for the root `file.txt`, you would need to add `!/file.txt` to whitelist it.

In general, whitelists like this are a bit of a pain. If you can manage it, it's much better to just say what you don't want to search.

**Reply by professionalPeter** [reply]
Thank you — such a huge help! I had tried both exceptions individually but didn't realize I'd need to use them together.

And FWIW, I agree that this is likely to be more of a pain than using the ignore file as intended. It's nothing about ripgrep or its documentation that led me down this road, but rather factors related to using this in the context of a team (e.g. trying to get similar results to teammates using other tools that define "include paths" rather than using an ignore file)
